File: plugins/sdd-harness/hooks/_sdd_common.py
"""Shared helpers for the global SDD/test hooks. Project- and OS-agnostic.

A hook is a no-op unless the project actually uses the workflow it guards:
- SDD hooks act only when `openspec/` is present (an OpenSpec project).
- "Active change" = the NEWEST change (by `created:` in its `.openspec.yaml`,
  tie-broken by dir name) still has an incomplete task (`- [ ]`). Pure
  dir-presence is useless because OpenSpec never moves finished changes out of
  `openspec/changes/`, and mtime is unreliable (git checkout/reset rewrites it,
  and stale changes keep stray unchecked closing tasks). Anchoring on the
  newest change means: once its tasks are all done you must propose a new
  change before touching production source again.

Global kill-switch: set SDD_HOOKS_DISABLED=1 to make every hook a no-op.
"""
import glob
import os
import shutil
import subprocess


def disabled() -> bool:
    return os.environ.get("SDD_HOOKS_DISABLED", "") not in ("", "0", "false", "no")


def auto_init_enabled() -> bool:
    """Auto-init defaults ON (the user wants every project spec-driven). Set
    SDD_AUTO_INIT=0 to keep the gate but stop it scaffolding openspec/ in new repos."""
    return os.environ.get("SDD_AUTO_INIT", "1") not in ("", "0", "false", "no")


def ensure_openspec_project(cwd: str) -> bool:
    """Make `cwd` an OpenSpec project if it is a git repo that isn't one yet, via a
    NON-INTERACTIVE `openspec init --tools none --force`. Returns True if it is (or just
    became) an OpenSpec project. Fail-open: any problem (auto-init off, not a git repo, no
    CLI on PATH, init error/timeout) returns False so the caller no-ops — never blocks an
    edit because setup failed. Idempotent: a no-op once openspec/ exists."""
    if is_openspec_project(cwd):
        return True
    if not auto_init_enabled():
        return False
    if not os.path.isdir(os.path.join(cwd, ".git")):
        return False
    exe = shutil.which("openspec")
    if not exe:
        return False
    try:
        subprocess.run(
            [exe, "init", "--tools", "none", "--force", cwd],
            cwd=cwd,
            capture_output=True,
            timeout=30,
            stdin=subprocess.DEVNULL,
        )
    except Exception:
        return False
    return is_openspec_project(cwd)


def is_openspec_project(cwd: str) -> bool:
    return os.path.isdir(os.path.join(cwd, "openspec", "changes")) or os.path.isfile(
        os.path.join(cwd, "openspec", "config.yaml")
    )


def _read_created(change_dir: str) -> str:
    for fn in (".openspec.yaml", ".openspec.yml"):
        path = os.path.join(change_dir, fn)
        if not os.path.isfile(path):
            continue
        try:
            with open(path, encoding="utf-8") as fh:
                for line in fh:
                    stripped = line.strip()
                    if stripped.startswith("created:"):
                        return stripped.split(":", 1)[1].strip()
        except OSError:
            return ""
    return ""


def has_active_change(cwd: str) -> bool:
    candidates = []
    for tasks in glob.glob(os.path.join(cwd, "openspec", "changes", "*", "tasks.md")):
        change_dir = os.path.dirname(tasks)
        name = os.path.basename(change_dir)
        if name == "archive":
            continue
        candidates.append((_read_created(change_dir), name, tasks))
    if not candidates:
        return False
    _, _, newest_tasks = max(candidates, key=lambda c: (c[0], c[1]))
    try:
        with open(newest_tasks, encoding="utf-8") as fh:
            return "- [ ]" in fh.read()
    except OSError:
        return False

File: plugins/sdd-harness/hooks/_sdd_common.mjs
// Shared helpers for the SDD/test hooks. Project- and OS-agnostic. Node built-ins only.
//
// A hook is a no-op unless the project actually uses the workflow it guards:
// - SDD hooks act only when `openspec/` is present (an OpenSpec project).
// - "Active change" = the NEWEST change (by `created:` in its `.openspec.yaml`, tie-broken
//   by dir name) still has an incomplete task (`- [ ]`). Pure dir-presence is useless
//   (OpenSpec never moves finished changes out of `openspec/changes/`) and mtime is
//   unreliable. Anchoring on the newest change means: once its tasks are all done you must
//   propose a new change before touching production source again.
//
// Global kill-switch: set SDD_HOOKS_DISABLED=1 to make every hook a no-op.
import fs from 'node:fs'
import path from 'node:path'
import { spawnSync } from 'node:child_process'

export function disabled() {
  return !['', '0', 'false', 'no'].includes(process.env.SDD_HOOKS_DISABLED || '')
}

// Auto-init defaults ON (every project spec-driven). SDD_AUTO_INIT=0 keeps the gate but
// stops it scaffolding openspec/ in new repos.
export function autoInitEnabled() {
  return !['', '0', 'false', 'no'].includes(process.env.SDD_AUTO_INIT ?? '1')
}

// Read ALL of stdin synchronously. Hooks receive their JSON via a pipe, and
// fs.readFileSync(0) returns '' for a pipe (it trusts the fstat size, which is 0), so we
// loop over readSync until EOF. Never throws.
export function readStdin() {
  const chunks = []
  const buf = Buffer.alloc(65536)
  while (true) {
    let n
    try { n = fs.readSync(0, buf, 0, buf.length, null) }
    catch (e) { if (e.code === 'EAGAIN') continue; break }
    if (!n) break
    chunks.push(Buffer.from(buf.subarray(0, n)))
  }
  return Buffer.concat(chunks).toString('utf8')
}

function isDir(p) { try { return fs.statSync(p).isDirectory() } catch { return false } }
function isFile(p) { try { return fs.statSync(p).isFile() } catch { return false } }

// Run an external tool cross-platform: on Windows a CLI installed by npm (openspec, pnpm,
// yarn, npm) is a `.cmd`, so try the common wrappers before giving up. Never throws.
export function spawnTool(name, args, opts = {}) {
  const names = process.platform === 'win32' ? [`${name}.cmd`, `${name}.exe`, name] : [name]
  for (const n of names) {
    const r = spawnSync(n, args, { encoding: 'utf8', ...opts })
    if (!r.error || r.error.code !== 'ENOENT') return r
  }
  return { error: new Error('ENOENT'), status: null, stdout: '', stderr: '' }
}

export function isOpenspecProject(cwd) {
  return isDir(path.join(cwd, 'openspec', 'changes')) || isFile(path.join(cwd, 'openspec', 'config.yaml'))
}

// Make `cwd` an OpenSpec project if it's a git repo that isn't one yet, via a
// non-interactive `openspec init --tools none --force`. Fail-open: any problem (auto-init
// off, not a git repo, no CLI, init error) returns false so the caller no-ops — never
// blocks an edit because setup failed. Idempotent once openspec/ exists.
export function ensureOpenspecProject(cwd) {
  if (isOpenspecProject(cwd)) return true
  if (!autoInitEnabled()) return false
  if (!isDir(path.join(cwd, '.git'))) return false
  const r = spawnTool('openspec', ['init', '--tools', 'none', '--force', cwd], { cwd, timeout: 30000, stdio: 'ignore' })
  if (r.error) return false
  return isOpenspecProject(cwd)
}

function readCreated(changeDir) {
  for (const fn of ['.openspec.yaml', '.openspec.yml']) {
    try {
      for (const line of fs.readFileSync(path.join(changeDir, fn), 'utf8').split('\n')) {
        const s = line.trim()
        if (s.startsWith('created:')) return s.slice('created:'.length).trim()
      }
    } catch { /* try next filename */ }
  }
  return ''
}

export function hasActiveChange(cwd) {
  const changesDir = path.join(cwd, 'openspec', 'changes')
  let entries
  try { entries = fs.readdirSync(changesDir, { withFileTypes: true }) } catch { return false }
  const candidates = []
  for (const e of entries) {
    if (!e.isDirectory() || e.name === 'archive') continue
    const tasks = path.join(changesDir, e.name, 'tasks.md')
    if (!isFile(tasks)) continue
    candidates.push([readCreated(path.join(changesDir, e.name)), e.name, tasks])
  }
  if (!candidates.length) return false
  // newest = max by (created, name), matching the Python tuple sort
  candidates.sort((a, b) => (a[0] < b[0] ? -1 : a[0] > b[0] ? 1 : a[1] < b[1] ? -1 : a[1] > b[1] ? 1 : 0))
  const newestTasks = candidates[candidates.length - 1][2]
  try { return fs.readFileSync(newestTasks, 'utf8').includes('- [ ]') } catch { return false }
}
